File: Day82/optimizedSolution.java

class Node {
    int data;
    Node left;
    Node right;
    Node(int data) {
        this.data = data;
        left = null;
        right = null;
    }
}

class Solution {
    
    int[] getHeightDiameter(Node root) {
        /**
         * This function recursivesly finds the height and diameter
         * of a binary tree
         * 
         * Approach:
         * 1) The height is the max of left and right nodes
         * 2) The diameter is the max of left, right and sum of height
         * 
         * Time Complexity: O(n)
         * Space Complexity: O(h)
         **/

        if(root == null)
            return new int[] {0, 0};
        
        int leftH, leftD, rightH, rightD;
        leftH = leftD = rightH = rightD = 0;
        
        if(root.left != null) {
            int[] leftHD = getHeightDiameter(root.left);
            leftH = 1 + leftHD[0];
            leftD = leftHD[1];
        }
            
        if(root.right != null) {
            int[] rightHD = getHeightDiameter(root.right);
            rightH = 1 + rightHD[0];
            rightD = rightHD[1];
        }
        
        int D = leftH + rightH, d = leftD > rightD ? leftD : rightD;
        
        int[] res = {
                leftH > rightH ? leftH : rightH,
                D > d ? D : d
            };
            
        return res;

    }
    
    int diameter(Node root) {
        // This function calls the above and returns the diameter
        
        int[] res = getHeightDiameter(root);
        
        return res[1];
        
    }
}

File: Day88/optimizedSolution.py

class Node:
    def __init__(self, val):
        self.right = None
        self.data = val
        self.left = None

class Solution:
    
    def treeSums(self, root, k, prefSums, currSum):
        
        if(root == None):
            return 0
        
        count = 0
        
        currSum += root.data
        
        if(currSum == k):
            count += 1
        
        count += prefSums.get(currSum - k, 0)
        
        prefSums[currSum] = prefSums.get(currSum, 0) + 1

        count += self.treeSums(root.left, k, prefSums, currSum)
        count += self.treeSums(root.right, k, prefSums, currSum)
        
        prefSums[currSum] = prefSums.get(currSum) - 1
        
        return count
    
    def sumK(self,root,k):
        
        prefSums = dict()
        
        count = self.treeSums(root, k, prefSums, 0)
        
        return count
